cdk/dcm_floodlight/transform.js

```diff
@@ -0,0 +1,171 @@
+/* eslint-disable no-param-reassign */
+const get = require("get-value");
+const {
+  getValueFromMessage,
+  removeUndefinedAndNullValues,
+  isDefinedAndNotNull
+} = require("rudder-transformer-cdk/build/utils");
+const { CustomError, getIntegrationsObj, isEmpty } = require("../../v0/util");
+
+// append properties to endpoint
+// eg: ${endpoint}key1=value1;key2=value2;....
+function appendProperties(endpoint, payload) {
+  Object.keys(payload).forEach(key => {
+    endpoint += `${key}=${payload[key]};`;
+  });
+
+  return endpoint;
+}
+
+// transform webapp dynamicForm custom floodlight variable
+// into {u1: value, u2: value, ...}
+function transformCustomVariable(customFloodlightVariable) {
+  const customVariable = {};
+  customFloodlightVariable.forEach(item => {
+    if (!isEmpty(item.from) && !isEmpty(item.to)) {
+      // remove u if already there
+      customVariable[
+        `u${item.from.trim().replace(/u/g, "")}`
+      ] = `[${item.to.trim()}]`;
+    }
+  });
+
+  return customVariable;
+}
+
+// valid flag should be provided [1|true] or [0|false]
+function isValidFlag(key, value) {
+  if (["true", "1"].includes(value.toString())) {
+    return 1;
+  }
+  if (["false", "0"].includes(value.toString())) {
+    return 0;
+  }
+
+  throw new CustomError(
+    `[DCM Floodlight]:: ${key}: valid parameters are [1|true] or [0|false]`,
+    400
+  );
+}
+
+function trackPostMapper(input, mappedPayload, rudderContext) {
+  const { message, destination } = input;
+  const { advertiserId, conversionEvents } = destination.Config;
+  let { activityTag, groupTag } = destination.Config;
+  let customFloodlightVariable;
+  let salesTag;
+
+  const baseEndpoint = "https://ad.doubleclick.net/ddm/activity/";
+
+  let event = getValueFromMessage(message, "event");
+  if (!event) {
+    throw new CustomError(
+      "[DCM Floodlight]:: event is required for track call",
+      400
+    );
+  }
+
+  const userAgent = get(message, "context.userAgent");
+  if (!userAgent) {
+    throw new CustomError(
+      "[DCM Floodlight]:: userAgent is required for track call",
+      400
+    );
+  }
+  rudderContext.userAgent = userAgent;
+
+  // find conversion event
+  // some() stops execution if at least one condition is passed and returns bool
+  event = event.trim().toLowerCase();
+  const conversionEventFound = conversionEvents.some(conversionEvent => {
```

Comment:
```suggestion
  const conversionEventFound = conversionEvents.some(conversionEvent => {
  if(conversionEvent) { ..
```
If conversionEvent exisits do the nextsteps, 
`conversionEvents = [undefined, null, etc]`
This can be a possibility considering control plane as black box

Reply:
`some` won't traverse if a particular conversion event does not exist.

Reply:
![Screenshot 2022-04-13 at 1 52 09 PM](https://user-images.githubusercontent.com/25389167/163132895-6d411fd5-b916-4aad-a62b-b9d467551cf5.png)
Seeing its iteration actually

Reply:
dynamicCustomForm takes care of this